### PR TITLE
feature: add option for including shaders on model export.

### PR DESCRIPTION
**Pull request by nikshankarnoble**
## Changelog Description

This is a duplication of the old PR https://github.com/ynput/ayon-core/pull/444 which got closed as part of the migration to the new repository. 

The current default behaviour when exporting a model is that the shaders from Maya will be stripped, but for certain workflows (such as ours) we want to pass the shaders down from the modelling step - this PR allows for that without changing the default behaviour from current.

## Additional info

@antirotor based on the discussion in the last PR and personal preference I switched the setting from `strip_shaders` to `include_shaders` and made the default False - let me know if you'd prefer it as a strip setting and I can update the PR.

## Testing notes:

Tested by exporting a model through Maya and ensuring that the model retains it's shaders with the option on - with existing behaviour when published without `include_shaders` on.

### CODE REVIEW

File: client/ayon_maya/plugins/create/create_model.py [hunk 1]
```diff
@@ -39,5 +39,9 @@ def get_instance_attr_defs(self):
                     placeholder="attr1, attr2"),
             TextDef("attrPrefix",
                     label="Custom Attributes Prefix",
-                    placeholder="prefix1, prefix2")
+                    placeholder="prefix1, prefix2"),
+            BoolDef("include_shaders",
```
Comment (nikshankarnoble) [inline client/ayon_maya/plugins/create/create_model.py:43]: @antirotor should this be `"include_shaders"` or `"includeShaders"`? I stuck with snake case to match your PR, but it seems that camelCase is the convention for the other settings.

Comment (BigRoy) [reply client/ayon_maya/plugins/create/create_model.py:43]: In this case the snake case is good. Thanks for checking! :)

**Review by LiborBatek — APPROVED** [review]
Default state been correctly as ```Disabled``` which is preferable.

Screenshot 2024-08-19 164924

All working fine for both states of ```Include shaders``` settings...

When disabled, only mesh without shaders been produced.

When enabled, shaders are included in model with suitable representation like maya ascii or FBX.

Screenshot 2024-08-19 165822

and when the ```model``` product been loaded and shader present too:

Screenshot 2024-08-19 165748

File: client/ayon_maya/plugins/create/create_model.py [hunk 2]
```diff
@@ -39,5 +39,9 @@ def get_instance_attr_defs(self):
...
+                    label="Include Shaders",
+                    tooltip="Include shaders in the geometry export.",
+                    default=False),
```
Comment (BigRoy) [inline client/ayon_maya/plugins/create/create_model.py:46]: @LiborBatek looking at this code I have a feeling the setting in settings does not influence the default state here at all?

Comment (BigRoy) [reply client/ayon_maya/plugins/create/create_model.py:46]: It should take it from a class attribute `self.include_shaders` similar to this: https://github.com/ynput/ayon-maya/pull/69/files#diff-d3f80c6626121c1d853dcc382eb66d18761a2aad4b6d7028125857fa5e552befR30

Of course `include_shaders = False` should also be added as class attribute.

```suggestion
                    default=self.include_shaders),
```

Comment (antirotor) [reply client/ayon_maya/plugins/create/create_model.py:46]: I agree - @nikshankarnoble I tried to fix it but I don't have permissions to push to your fork. Could you just make those changes? With that, it is good to be merged. Thanks!

**Review by antirotor — DISMISSED** [review]
apart that default value, it looks great

**Comment by LiborBatek** [comment]
@BigRoy @antirotor yep, it was the case...I had to enable it manually (but as I was testing with already existing ```model``` instance I was thinking it was in its default disabled state due to the fact it was created before the option been available aka in this code branch) - I wasnt aware of this malfunction then...

Great catch!

File: client/ayon_maya/plugins/create/create_model.py (hunk 2 above)
Comment (BigRoy) [reply client/ayon_maya/plugins/create/create_model.py]: I was able to push. ✅

**Comment by BigRoy** [comment]
> @BigRoy @antirotor yep, it was the case...I had to enable it manually (but as I was testing with already existing `model` instance I was thinking it was in its default disabled state due to the fact it was created before the option been available aka in this code branch) - I wasnt aware of this malfunction then...
> 
> Great catch!

@LiborBatek could you try again with latest commit?

**Review by LiborBatek — CHANGES_REQUESTED** [review]
Works well (Include Shader ON) for ```.ma``` , ```FBX```

but not working with ```OBJ``` repre type....it lacks the shader assignments

image


note: when creating fresh new ```model``` publish instances the ```Include shader``` and its state correctly get propagated (enabled/disabled)...so all good on this front

File: client/ayon_maya/plugins/create/create_model.py (hunk 2 above)
Comment (nikshankarnoble) [reply client/ayon_maya/plugins/create/create_model.py]: Lovely - thanks very much both!

**Comment by LiborBatek** [comment]
How about that ```OBJ``` and included shaders on?? Are we resolving it? As they arent included when publish happens...

Otherwise all working nicely!

**Comment by antirotor** [comment]
> How about that `OBJ` and included shaders on?? Are we resolving it? As they arent included when publish happens...
> 
> Otherwise all working nicely!

If I remember correctly from the previous PR it didn't work for OBJ even when you are setting the options. Might be bug in OBJ exporter in Maya

**Comment by BigRoy** [comment]
@LiborBatek this took me way longer to debug why it didn't work until I realized in `.obj` the materials are stored in a separate `.mtl` file on export. So the export does work - but the integrated files do not work for `.obj`.

I'd say - it's a bit out of scope for this PR to make it work - but the issue is essentially:

1. We need to also integrate the `.mtl` file for the obj.
2. We need to ALSO change contents in the `.obj` file because it'd store e.g. a line like: `mtllib modelMain.mtl` which would then need to be changed to the published integrated filename for the `.mtl` file (unless we transfer the file without renaming, almost like a 'resource' and make sure it lives next to the `.obj` then it'd work).

**Comment by mkolar** [comment]
I will play devils advocate here and ask....shouldn't we finally deprecated .obj in the default completely? is that even the right format to use nowadays?

**Comment by LiborBatek** [comment]
```OBJ``` stands for something like "if any other method fails" use it! Its quite common to bring OBJ with material groups into Substance Painter, ZBrush etc. So I think we should keep it (on a pretty basic level)...

@BigRoy I would go with ```mat``` file as a resource file without renaming when integrate it...making things much easier! and still working as expected

**Comment by BigRoy** [comment]
> I would go with mat file as a resource file without renaming when integrate it...making things much easier! and still working as expected

Fine with me - but.... this PR or separate PR?

**Comment by BigRoy** [comment]
> I would go with `mat` file as a resource file without renaming when integrate it...making things much easier! and still working as expected

Added to this PR.
@LiborBatek please test again.

**Review by LiborBatek — CHANGES_REQUESTED** [review]
Now it does produce ```OBJ``` with accompanying ```mat``` file but it lacks texure connections (not sure if we can tackle it easily)

here is my ```OBJ``` model loaded (aka imported) into clean scene and also inspected resources:
Screenshot 2024-08-28 150841

published repre:

Screenshot 2024-08-28 150859

and when inspecting ```mat``` file in text editor:

Screenshot 2024-08-28 150915

**Comment by BigRoy** [comment]
> file but it lacks texure connections (not sure if we can tackle it easily)

Does this work with regular `.obj` export for you? This may be a limitation of the `.obj` format for certain shaders. Can you confirm whether this is due to how we export it for AYON or just happens to be an issue with OBJ in general? @LiborBatek

**Comment by LiborBatek** [comment]
Small note: I also lack the ```OBJ``` extractor knob when set in Ayon Settings as ```Optional``` it is not just present in the ```Publisher``` UI...

image

**Comment by BigRoy** [comment]
> Small note: I also lack the `OBJ` extractor knob when set in Ayon Settings as `Optional` it is not just present in the `Publisher` UI...
> 
> image

Fixed with https://github.com/ynput/ayon-maya/pull/69/commits/f04a0c26715d146bd5dfd8e440af368060e9c77d

**Review by LiborBatek — CHANGES_REQUESTED** [review]
Now it correctly shows ```Extract OBJ``` as optional within Publisher UI..so this been fixed!

Screenshot 2024-08-28 153933

Now also the ```mat``` file contains the textures inside...but we miss the actual image files copied next to the ```OBJ``` file...as when doing so manually (File>Export selection (OBJ))

Screenshot 2024-08-28 154119

here is the result when done manually:



Screenshot 2024-08-28 153230

**Comment by BigRoy** [comment]
I really am starting to hate the `.obj` and `.mtl` file format.
Please try again. :)

**Review by LiborBatek — APPROVED** [review]
Now all works as expected and also ```OBJ``` representation having shader and corresponding source files / textures included.

All good! I think its good to go!

Screenshot 2024-08-29 095812

Also tested by importing the ```OBJ``` into empty workfile...

**Comment by BigRoy** [comment]
@mkolar will hate me - but I'm merging this :) We can fight later.
